### PR TITLE
feat: parse prs

### PR DESCRIPTION
**Pull request by MounirDhahri**
The type of this PR is: **feature**

This PR resolves [CX-1213]

### Description
This PR adds a script that generates and adds changelog changes to a platform specific changelog file.
`./scripts/changelog/generatePlatformChangelog -p android` or `./scripts/changelog/generatePlatformChangelog -p ios` should generate changelog files with the changes since the last promoted release. See video

https://user-images.githubusercontent.com/11945712/120814685-67d48000-c54f-11eb-9fac-638b42b568d1.mov

### PR Checklist (tick all before merging)

- [x] I have included screenshots or videos to illustrate my changes, or I have not changed anything that impacts the UI.
- [x] I have tested my changes on **iOS** and **Android**.
- [x] I have added tests for my changes, or my changes don't require testing, or I have included a link to a separate Jira ticket covering the tests.
- [x] I have added a feature flag, or my changes don't require a feature flag. ([How do I add one?](https://github.com/artsy/eigen/blob/master/docs/developing_a_feature.md))
- [x] I have documented any follow-up work that this PR will require, or it does not require any.
- [x] I have added an app state migration, or my changes do not require one. ([What are migrations?](https://github.com/artsy/eigen/blob/master/docs/adding_state_migrations.md))
- [x] I have added a [CHANGELOG.yml](/CHANGELOG.yml) entry or my changes do not require one.

[CX-434]: https://artsyproduct.atlassian.net/browse/CX-434

[CX-1213]: https://artsyproduct.atlassian.net/browse/CX-1213

### CODE REVIEW

**Comment by MounirDhahri** [comment]
I just found out I didn't assign anyone to review this 😔 sorry about that

File: fastlane/Fastfile [hunk 1]
```diff
@@ -409,6 +409,10 @@ lane :promote_beta_android do
     skip_upload_images: true,
     skip_upload_screenshots: true,
   )
+
+  vname, vcode = set_build_version_android
+  tag_and_push(tag: "android-#{vname}-#{vcode}-submission")
+
```
Comment (MounirDhahri) [inline fastlane/Fastfile:415]: I added tagging to our android submission releases

File: scripts/changelog/generatePlatformChangelog.js [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+    iOSUserFacingChanges: [],
+  }
+
+  prs
+    .map((pr) => parsePRDescription(pr.body))
+    .filter((parseResult) => parseResult.type === "changes")
+    .map((parseResult) => {
+      for (const entry of Object.keys(changeLog)) {
+        // @ts-ignore
+        changeLog[entry] = changeLog[entry].concat(parseResult[entry])
```
Comment (MounirDhahri) [inline scripts/changelog/generatePlatformChangelog.js:132]: Typescript was really good at parsing a js file until it failed here to know that entry is actually the right type 😅

File: scripts/changelog/parsePRDescription.js [hunk 3]
```diff
@@ -8,7 +8,7 @@ const { changelogTemplateSections } = require("./generateChangelogSectionTemplat
  */
 module.exports.parsePRDescription = (description) => {
   description = stripComments(description)
-  if (description.includes("#nochangelog")) {
+  if (description.includes("#nochangelog") || !description.includes("#run_new_changelog_check")) {
```
Comment (MounirDhahri) [inline scripts/changelog/parsePRDescription.js:11]: I am checking for #run_new_changelog_check to make sure we don't break anything once we deploy this. I will of course remove this once we finish everything

File: scripts/changelog/generatePlatformChangelog-tests.ts [hunk 4]
```diff
@@ -0,0 +1,155 @@
+import { getCombinedChangeLog, getPlafromSpecificChangeLog, isMergedAfter } from "./generatePlatformChangelog"
```
Comment (brainbicycle) [inline scripts/changelog/generatePlatformChangelog-tests.ts:1]:
```suggestion
import { getCombinedChangeLog, getPlaformSpecificChangeLog, isMergedAfter } from "./generatePlatformChangelog"
```
minor typo here

File: scripts/changelog/generatePlatformChangelog-tests.ts [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+describe("isMergedAfter", () => {
+  it("returns false when the PR has not been merged (merged_at is null)", () => {
+    expect(isMergedAfter(null, new Date("01 June 2021"))).toEqual(false)
+  })
+
+  it("returns false when the PR has not been merged before the release commit date", () => {
+    expect(isMergedAfter("01 June 2021 ", new Date("02 June 2021"))).toEqual(false)
+  })
+
+  it("returns false when the PR has been merged after the release commit date ", () => {
```
Comment (brainbicycle) [inline scripts/changelog/generatePlatformChangelog-tests.ts:29]:
```suggestion
  it("returns true when the PR has been merged after the release commit date ", () => {
```

File: scripts/changelog/generatePlatformChangelog-tests.ts [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    expect(getCombinedChangeLog([] as any)).toEqual({
+      androidUserFacingChanges: [],
+      crossPlatformUserFacingChanges: [],
+      devChanges: [],
+      iOSUserFacingChanges: [],
+    })
+  })
+})
+
+describe("getPlafromSpecificChangeLog", () => {
```
Comment (brainbicycle) [inline scripts/changelog/generatePlatformChangelog-tests.ts:53]:
```suggestion
describe("getPlaformSpecificChangeLog", () => {
```
minor typo :)

File: scripts/changelog/generatePlatformChangelog-tests.ts [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    - Added an other button
+    - Fixed modal close button
+    - fixed button on iOS
+
+  - Dev changes:
+    - Improved changelog tooling
+    - Upgraded lodash
+`
+    )
+  })
```
Comment (brainbicycle) [inline scripts/changelog/generatePlatformChangelog-tests.ts:82]: Nice tests! 🙌

File: scripts/changelog/generatePlatformChangelog.js [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+  - Build tag: **${commitTag}**
+  - App store submission date: **Fri Jun 04 2021 15:48:24 GMT+0200 (Central European Summer Time)**
+  - Changelog:
+    - User facing changes:
+      - Made button bigger
+    - Dev changes:
+      - Fixed rerendering issues
+ *
+ */
+function getPlafromSpecificChangeLog(platform, changelog) {
```
Comment (brainbicycle) [inline scripts/changelog/generatePlatformChangelog.js:181]:
```suggestion
function getPlaformSpecificChangeLog(platform, changelog) {
```

**Review by brainbicycle — APPROVED** [review]
Fantastic work! Very excited to get this in, a few minor typos in function name but code looks great and great tests! Once typos are fixed feel free to merge 👍

File: docs/pull_request_template.md [hunk 9]
```diff
@@ -22,3 +22,29 @@ This PR resolves [CX-]
...
+
+### Changelog updates
+
+<!-- 📝 Please fill out at least one of these sections. -->
+<!-- ⓘ 'User-facing' changes will be published as release notes. -->
+<!-- ⌫ Feel free to remove sections that don't apply. -->
+<!-- • Write a markdown list or just a single paragraph, but stick to plain text. -->
+<!-- 🤷‍♂️ Replace this entire block with the hashtag `#nochangelog` to avoid updating the changelog. -->
+
+#### Android user-facing changes
```
Comment (olerichter00) [inline docs/pull_request_template.md:34]: (non-blocking) Can we sort them a bit?  Intuitively I would sort them like this:
- Cross-platform
- iOS
- Android
- Dev

File: CHANGELOG/android-changelog.md [hunk 10]
```diff
@@ -0,0 +1,7 @@
+## Undeployed Changes
+
+-
+
+<!-- DO NOT CHANGE -->
+
+## Deployed Changes
```
Comment (olerichter00) [inline CHANGELOG/android-changelog.md:7]: (non-blocking) Wouldn't `Released Changes` make more sense?

File: docs/pull_request_template.md [hunk 11]
```diff
@@ -22,3 +22,29 @@ This PR resolves [CX-]
...
 - [ ] I have added an app state migration, or my changes do not require one. ([What are migrations?](https://github.com/artsy/eigen/blob/master/docs/adding_state_migrations.md))
 - [ ] I have added a [CHANGELOG.yml](/CHANGELOG.yml) entry or my changes do not require one.
+
+### Changelog updates
+
+<!-- 📝 Please fill out at least one of these sections. -->
+<!-- ⓘ 'User-facing' changes will be published as release notes. -->
+<!-- ⌫ Feel free to remove sections that don't apply. -->
+<!-- • Write a markdown list or just a single paragraph, but stick to plain text. -->
+<!-- 🤷‍♂️ Replace this entire block with the hashtag `#nochangelog` to avoid updating the changelog. -->
```
Comment (olerichter00) [inline docs/pull_request_template.md:32]: (question) Would it break the flow to not write `#nochangelog`?

File: scripts/changelog/generatePlatformChangelog-tests.ts (hunk 4 above)
Comment (olerichter00) [reply scripts/changelog/generatePlatformChangelog-tests.ts:1]:
```suggestion
import { getCombinedChangeLog, getPlatfromSpecificChangeLog, isMergedAfter } from "./generatePlatformChangelog"
```

File: scripts/changelog/generatePlatformChangelog-tests.ts (hunk 7 above)
Comment (olerichter00) [reply scripts/changelog/generatePlatformChangelog-tests.ts:82]: So many buttons!

File: scripts/changelog/generatePlatformChangelog.js [hunk 12]
```diff
@@ -0,0 +1,242 @@
+#!/usr/bin/env node
+
+// @ts-check
```
Comment (olerichter00) [inline scripts/changelog/generatePlatformChangelog.js:3]: 🎉

File: scripts/changelog/generatePlatformChangelog.js [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+ * @param {"android" | "ios"} platform
+ * @param {import("./changelog-types").ParseResultChanges} changelog
+ */
+function generatePlatformChangelog(platform, changelog) {
+  const spinner = ora("Generating platform specific changelog")
+
+  const changelogFilePath = `./CHANGELOG/${platform}-changelog.md`
+  let fileContent = fs.readFileSync(changelogFilePath, "utf8").toString()
+
+  const regex = /## Deployed Changes/
```
Comment (olerichter00) [inline scripts/changelog/generatePlatformChangelog.js:149]: Same here. Wouldn't `Released Changes` make more sense?

**Review by olerichter00 — APPROVED** [review]
Looks amazing!

File: docs/pull_request_template.md (hunk 11 above)
Comment (MounirDhahri) [reply docs/pull_request_template.md:32]: no, it should not break the flow. it will just skip the check

File: docs/pull_request_template.md (hunk 9 above)
Comment (MounirDhahri) [reply docs/pull_request_template.md]: yes, that makes sense

File: scripts/changelog/generatePlatformChangelog-tests.ts (hunk 4 above)
Comment (MounirDhahri) [reply scripts/changelog/generatePlatformChangelog-tests.ts]: thanks

File: scripts/changelog/generatePlatformChangelog-tests.ts (hunk 6 above)
Comment (MounirDhahri) [reply scripts/changelog/generatePlatformChangelog-tests.ts]: good catch

**Pull request merged**